pkg/reconciler/inmemorychannel/inmemorychannel.go

```diff
@@ -0,0 +1,295 @@
+}
+
+func (r *Reconciler) OnUpdate(old, new interface{}) {
+	r.impl.GlobalResync(r.inmemorychannelInformer)
+}
+
+func (r *Reconciler) OnDelete(obj interface{}) {
+	r.impl.GlobalResync(r.inmemorychannelInformer)
+}
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the InMemoryChannel resource
+// with the current status of the resource.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	// Convert the namespace/name string into a distinct namespace and name.
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logging.FromContext(ctx).Error("invalid resource key")
+		return nil
+	}
+
+	// Get the InMemoryChannel resource with this namespace/name.
+	original, err := r.inmemorychannelLister.InMemoryChannels(namespace).Get(name)
+	if apierrs.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logging.FromContext(ctx).Error("InMemoryChannel key in work queue no longer exists")
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy.
+	channel := original.DeepCopy()
+
+	// Reconcile this copy of the InMemoryChannel and then write back any status updates regardless of
+	// whether the reconcile error out.
+	reconcileErr := r.reconcile(ctx, channel)
+	if reconcileErr != nil {
+		logging.FromContext(ctx).Error("Error reconciling InMemoryChannel", zap.Error(reconcileErr))
+		r.Recorder.Eventf(channel, corev1.EventTypeWarning, channelReconcileFailed, "InMemoryChannel reconciliation failed: %v", reconcileErr)
+	} else {
+		logging.FromContext(ctx).Debug("InMemoryChannel reconciled")
+		r.Recorder.Event(channel, corev1.EventTypeNormal, channelReconciled, "InMemoryChannel reconciled")
+	}
+
+	if _, updateStatusErr := r.updateStatus(ctx, channel); updateStatusErr != nil {
+		logging.FromContext(ctx).Error("Failed to update InMemoryChannel status", zap.Error(updateStatusErr))
+		r.Recorder.Eventf(channel, corev1.EventTypeWarning, channelUpdateStatusFailed, "Failed to update InMemoryChannel's status: %v", err)
+		return updateStatusErr
+	}
+
+	// Requeue if the resource is not ready
+	return reconcileErr
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, imc *v1alpha1.InMemoryChannel) error {
+	imc.Status.InitializeConditions()
+
+	if imc.DeletionTimestamp != nil {
+		// Everything is cleaned up by the garbage collector.
+		return nil
+	}
+
+	// We reconcile the status of the Channel by looking at:
+	// 1. Dispatcher Deployment for it's readiness.
+	// 2. Dispatcher k8s Service for it's existence.
+	// 3. Dispatcher endpoints to ensure that there's something backing the Service.
+
+	// Get the Dispatcher Deployment and propagate the status to the Channel
+	d, err := r.deploymentLister.Deployments(r.dispatcherNamespace).Get(r.dispatcherDeploymentName)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
```

Comment:
Question: I have noticed a pattern where we create the dispatcher using the yaml and only monitor it here for status. Why don't we create it from the controller? So if the deployment get's deleted or if any of its dependencies such as the role get deleted, then the controller can recreate them and bring the cluster state back to desired state?
Not asking to make the change, just trying to understand the rationale.

Reply:
hahaha, I thought about that as well initially. Talked with @Harwayne and @mattmoor about this and decided that we'll just create it from the YAML and there will be operator that's going to be responsible for reconciling this. If we move to per namespaced or some other more granular dispatcher strategies, then we'll need to create these dynamically.

Reply:
Any particular reason we decided to go with yaml route or just to keep the implementation simple for now?

Reply:
simpler implementation.